### PR TITLE
Linux: reorder /proc/<pid>/status parsing

### PR DESCRIPTION
**Pull request by cgzones**
In case parsing an essential pid entry file like 'status' we treat the process as a short living one and ignore it.  In the relevant goto label the process structure is free'd, thus it must not have been inserted into the global process table.

Reorder parsing the status file after potentially inserting the process into the process table.

Fixes: 22d25db46 ("Linux: detect container process by different PID namespace")
Closes: #1455

### CODE REVIEW

File: linux/LinuxProcessTable.c [hunk 1]
```diff
@@ -1571,6 +1571,32 @@ static bool LinuxProcessTable_recurseProcTree(LinuxProcessTable* this, openat_ar
       if (!LinuxProcessTable_updateUser(host, proc, procFd, mainTask))
          goto errorReadingProcess;
 
+      /* Check if the process in inside a different PID namespace. */
```
Comment (BenBE) [inline linux/LinuxProcessTable.c:1574]:
```suggestion
      /* Check if the process is inside a different PID namespace. */
```

**Review by BenBE — APPROVED** [review]
Two minor nits …

File: linux/LinuxProcessTable.c [hunk 2]
```diff
@@ -1571,6 +1571,32 @@ static bool LinuxProcessTable_recurseProcTree(LinuxProcessTable* this, openat_ar
...
+      if (proc->isRunningInContainer == TRI_INITIAL && rootPidNs != (ino_t)-1) {
+         struct stat sb;
+#if defined(HAVE_OPENAT) && defined(HAVE_FSTATAT)
+         int res = fstatat(procFd, "ns/pid", &sb, 0);
+#else
+         char path[4096];
```
[the remark this answers is not in the document]
Comment (cgzones) [reply linux/LinuxProcessTable.c]: done

**Comment by Explorer09** [comment]
A minor gripe here: The commit description isn't obvious that it fixes a use-after-free. I would write the description like this:

```text
Linux: read pid status before inserting it into process table

When parsing an essential pid entry file like 'status' fails, we treat
the process as a short living one and skip adding it into the process
table. This step should be done before the process is added, or else
the goto label can free the process structure, creating use-after-free
scenarios later on.

Fixes: 22d25db ("Linux: detect container process by different PID namespace")
Closes: htop-dev#1455
```

**Pull request merged**